File: src/tools/file-helpers.ts
import { randomUUID } from "node:crypto";
import { link, open, rename, rm } from "node:fs/promises";
import path from "node:path";

export const MAX_FILE_WRITE_BYTES = 1024 * 1024;
export const MAX_FILE_READ_BYTES = 2 * 1024 * 1024;

export async function writeFileAtomic(
  target: string,
  content: string,
  mode: number,
  replace: boolean,
): Promise<void> {
  const directory = path.dirname(target);
  const temporary = path.join(
    directory,
    `.nash-write-${process.pid}-${randomUUID()}`,
  );
  const handle = await open(temporary, "wx", mode);
  try {
    try {
      await handle.writeFile(content, { encoding: "utf8" });
      await handle.sync();
    } finally {
      await handle.close();
    }
    if (replace) {
      await rename(temporary, target);
    } else {
      // Hard-linking a complete temporary file makes create-only writes
      // atomic and prevents a concurrent creator from being overwritten.
      await link(temporary, target);
    }
  } finally {
    await rm(temporary, { force: true });
  }
}

export function decodeUtf8(content: Uint8Array): string {
  try {
    return new TextDecoder("utf-8", { fatal: true }).decode(content);
  } catch {
    throw new InvalidEncodingError();
  }
}

export class InvalidEncodingError extends Error {
  public constructor() {
    super("file is not valid UTF-8");
    this.name = "InvalidEncodingError";
  }
}

export function errorCode(error: unknown): string | undefined {
  if (typeof error !== "object" || error === null || !("code" in error)) {
    return undefined;
  }
  return (error as NodeJS.ErrnoException).code;
}

export function messageOf(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

File: internal/tool/file_helpers.go
package tool

import (
	"fmt"
	"os"
	"path/filepath"
)

const maxFileWriteBytes = 1024 * 1024

func writeFileAtomic(path string, content []byte, mode os.FileMode, replace bool) error {
	directory := filepath.Dir(path)
	if err := os.MkdirAll(directory, 0o755); err != nil {
		return fmt.Errorf("create parent directory: %w", err)
	}

	temporary, err := os.CreateTemp(directory, ".nash-write-*")
	if err != nil {
		return fmt.Errorf("create temporary file: %w", err)
	}
	temporaryPath := temporary.Name()
	defer os.Remove(temporaryPath)

	if err := temporary.Chmod(mode.Perm()); err != nil {
		temporary.Close()
		return fmt.Errorf("set temporary file mode: %w", err)
	}
	if _, err := temporary.Write(content); err != nil {
		temporary.Close()
		return fmt.Errorf("write temporary file: %w", err)
	}
	if err := temporary.Sync(); err != nil {
		temporary.Close()
		return fmt.Errorf("sync temporary file: %w", err)
	}
	if err := temporary.Close(); err != nil {
		return fmt.Errorf("close temporary file: %w", err)
	}
	if replace {
		if err := os.Rename(temporaryPath, path); err != nil {
			return fmt.Errorf("replace target file: %w", err)
		}
		return nil
	}

	// Linking a complete temporary file gives create-only writes an atomic
	// no-clobber guarantee that a pre-write Stat check cannot provide.
	if err := os.Link(temporaryPath, path); err != nil {
		return fmt.Errorf("create target file: %w", err)
	}
	return nil
}
